`ProcessInstanceAssert.hasCalledProcess` doesn't check if the process contains the passed process id

Wolgo opened the pull request:
Actually check if a process with the correct processID is present

## Description

Improve function to actually test for correct process ID. 

## Related issues

<!-- Which issues are closed by this PR or are related -->

closes #

<!-- Cut-off marker
_All lines under and including the cut-off marker will be removed from the merge commit message_

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes
-->

## Definition of Done

<!-- Please check the items that apply, before merging or (if possible) before requesting a review. -->

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to backport the fix

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually

Documentation:
* [ ] Javadoc has been written
* [ ] The documentation is updated

Wolgo commented:
@remcowesterhoud I'm rudely tagging you, as I am otherwise going to forget about this, and it  Seems like a pretty important one, as this might result in tests incorrectly showing something calls the right subprocess.  Seems to me like this function was just forgetting to actually test for the correct ProcessId.

Didn't update tests or anything yet, because I have some other priorities right now.

remcowesterhoud reviewed (commented):
Thanks for the bugfix @Wolgo! Your changes look great, I like how you kept the old method available so we can stay backwards compatible 👍 

As you mentioned you haven't written tests. I can't merge this without tests, but I do want to see this merged. We can do 2 things here. If you are interested in writing the tests still you can commit this and I will merge this PR. Alternatively, I can create an issue and branch for this bug. I can merge it to this branch and write some tests for this myself. What do you prefer?

Wolgo commented:
`  public ProcessInstanceAssert findFirstCalledSubProcess(ProcessInstanceEvent ProcessEvent) {
    return new ProcessInstanceAssert(
        (StreamFilter.processInstance(this.recordStream)
            .withParentProcessInstanceKey(ProcessEvent.getProcessInstanceKey()).stream()
            .findFirst().orElseThrow(NoSuchElementException::new)).getKey(), this.recordStream);
  }`
  
  I quickly wrote this to work around that, and I'm thinking of just making it a generic function where you can get xth calls of process y.

remcowesterhoud commented:
> @remcowesterhoud I think it's good If I write some tests, do you have any examples? I saw these classes were not really doing any useful tests yet. I don't know if we wanna juggle with a mock setup or just use a real process.

It would be something along the lines of this in the `AbstractProcessInstanceAssertTest`:
```java
    @Test
    void testHasCalledProcess() throws InterruptedException, TimeoutException {
      // given
      Utilities.deployResources(
          client,
          ProcessPackCallActivity.RESOURCE_NAME,
          ProcessPackCallActivity.CALLED_RESOURCE_NAME);

      // when
      final ProcessInstanceEvent instanceEvent =
          Utilities.startProcessInstance(engine, client, ProcessPackCallActivity.PROCESS_ID);

      // then
      BpmnAssert.assertThat(instanceEvent)
          .hasCalledProcess(ProcessPackCallActivity.CALLED_PROCESS_ID);
    }
```

Of course this is only a happy path test for one of the functions. It's also not a great test as this is the only call activity that's in this process. So I would expect we need to introduce a `ProcessPackMultipleCallActivities` which contains a process with multiple call activities with different ids.

> I'm running to some other issues with subprocesses, so I'll probably extend it a bit more. (Multi-instances always seem to do X+1 calls, meaning that ExtractingLastCalledProcess becomes useless in the case of multi-instances.)

Would you mind separating this in a different branch? That way we can keep this PR isolated to one issue at a time.

remcowesterhoud commented:
Hi @Wolgo, hope you're doing well! This PR has been stale for a while, are you still interested in finishing this?

Wolgo commented:
Hey @remcowesterhoud , doing well, between a programming job and trying and succeeding a house, I never ended up finishing this.
Luckily you caught me on day where I could 0 coding and only did paperwork, so was pretty excited for this.
I wrote some tests for this, they are a bit wierd, as we can't check the processId of the current process in scope, so instead we use the fact that a process has or has not called a process to differentiate. (To me that's a bit easier to keep in mind, as that's already in scope, you could look at specific elements passed, but then you have to hold two concepts in mind.)

I am open for suggestions on alternatives that make more sense here. 
I added a small line about multi-instance loops, that's not a bug of this test code but maybe of camunda. Can't really do anything about that here, but it's a good warning for people :). 

I added a "HasNotCalledProcess" as that seems convenient in general and was really helpful for tests.

remcowesterhoud commented on qa/abstracts/src/main/java/io/camunda/zeebe/process/test/qa/abstracts/util/Utilities.java:
```diff
@@ -244,7 +244,19 @@ public static final class ProcessPackTimerStartEvent {
+    public static final String RESOURCE_NAME = "multiple-call-activity.bpmn";
+    public static final String PROCESS_ID = "multiple-call-activity";
+    public static final String CALL_ACTIVITY_ID_ONE = "callactivityOne";
+    public static final String CALLED_RESOURCE_NAME_ONE = ProcessPackAlternateStartEndEvent.RESOURCE_NAME;
+    public static final String CALLED_PROCESS_ID_ONE = ProcessPackAlternateStartEndEvent.PROCESS_ID;
```
❌ Instead of introducing a `ProcessPackAlternateStartEndEvent` let's just use the `ProcessPackStartEndEvent`. This means you can reference the resource from the `ProcessPackStartEndEvent`. For the process id you'll need some unique string, e.g. `calledProcessIdOne`.

Wolgo replied:
I need a different processId between One and Three, to show a hasCalled(OneId) but hasNotCalled(ThreeId) on the parent project. 

Some of the requirements for my process that I have in mind: 
- Two sequential call activities, to show it correctly extracts the last called process.
- A call activity that has a call activity, to show the parent process does not detect nested child processes. (Which also means that the nested child process needs to have a different process Id than the 2 non-nested child processes.)

The process id is baked into the resource, so I made a new simple resource to prevent having to do some extra step.
Is there a way to deploy the same process twice with a different process id? 
Or is there another process that we can use for this?

remcowesterhoud replied:
Oh sorry, my bad. I got confused with how we create processes in the Zeebe repo. You are totally right that you need the second process and resource. You can disregard these comments about removing the resource.

remcowesterhoud reviewed (changes requested):
> between a programming job and trying and succeeding to buy a house,

Sounds exciting!

Thanks for the changes you've made. Very thorough with the testing, I like it 🚀 

I have some minor suggestions. Mostly they are just some terminology fixes, but there's a few other changes as well. Please have a look and let me know if there's any questions!

Once you're done with the changes please run `mvn spotless:apply`. The CI is failing because of some code style issues 🙂

Wolgo commented:
I'll finish the rest this evening, my pipeline is done, so back to actual work

remcowesterhoud commented on qa/abstracts/src/main/resources/alternate-start-end.bpmn:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:modeler="http://camunda.org/schema/modeler/1.0" id="Definitions_03pj5sg" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion=" ...
+  <bpmn:process id="alternate-start-end" isExecutable="true">
+    <bpmn:startEvent id="StartEvent_1">
+      <bpmn:outgoing>Flow_0puuckb</bpmn:outgoing>
+    </bpmn:startEvent>
+    <bpmn:endEvent id="Event_0ahcnkn">
+      <bpmn:incoming>Flow_0puuckb</bpmn:incoming>
+    </bpmn:endEvent>
+    <bpmn:sequenceFlow id="Flow_0puuckb" sourceRef="StartEvent_1" targetRef="Event_0ahcnkn" />
+  </bpmn:process>
+  <bpmndi:BPMNDiagram id="BPMNDiagram_1">
+    <bpmndi:BPMNPlane id="BPMNPlane_1" bpmnElement="alternate-start-end">
+      <bpmndi:BPMNShape id="_BPMNShape_StartEvent_2" bpmnElement="StartEvent_1">
+        <dc:Bounds x="179" y="79" width="36" height="36" />
+      </bpmndi:BPMNShape>
+      <bpmndi:BPMNShape id="Event_0ahcnkn_di" bpmnElement="Event_0ahcnkn">
+        <dc:Bounds x="272" y="79" width="36" height="36" />
+      </bpmndi:BPMNShape>
+      <bpmndi:BPMNEdge id="Flow_0puuckb_di" bpmnElement="Flow_0puuckb">
+        <di:waypoint x="215" y="97" />
+        <di:waypoint x="272" y="97" />
+      </bpmndi:BPMNEdge>
+    </bpmndi:BPMNPlane>
+  </bpmndi:BPMNDiagram>
+</bpmn:definitions>
```
Please ignore

remcowesterhoud commented on qa/abstracts/src/main/java/io/camunda/zeebe/process/test/qa/abstracts/util/Utilities.java:
```diff
@@ -259,4 +271,9 @@ public static final class ProcessPackStartEndEvent {
     public static final String RESOURCE_NAME = "start-end.bpmn";
     public static final String PROCESS_ID = "start-end";
   }
+  public static final class ProcessPackAlternateStartEndEvent {
+
+    public static final String RESOURCE_NAME = "alternate-start-end.bpmn";
+    public static final String PROCESS_ID = "alternate-start-end";
+  }
```
Please ignore

Wolgo commented on assertions/src/main/java/io/camunda/zeebe/process/test/assertions/ProcessInstanceAssert.java:
```diff
@@ -593,18 +615,63 @@ public ProcessInstanceAssert extractingLatestCalledProcess(final String processI
+    final boolean hasCalledProcess = getCalledProcessRecords().stream().findAny().isPresent();
+
+    assertThat(hasCalledProcess)
+        .withFailMessage("A process was called from this process")
```
I didn't see a way to count how many times each process is called, and if you are not intending to call a process, being told which distinct processes are called feels pretty good already.
But agreed, so implemented.

remcowesterhoud commented on assertions/src/main/java/io/camunda/zeebe/process/test/assertions/ProcessInstanceAssert.java:
```diff
@@ -634,7 +634,9 @@ public ProcessInstanceAssert hasNotCalledProcess() {
-        .withFailMessage("A process was called from this process")
+        .withFailMessage("A process was called from this process, distinct called processes are: %s",
+            getCalledProcessRecords().stream().map(
+                x -> x.getValue().getBpmnProcessId()).distinct().collect(Collectors.toList()).toString())
```
🔧 One more minor change before I can merge it. We already get the caller process records on line 634. Please store the result of this in a variable instead of getting them again here. E.g.:

```java
  public ProcessInstanceAssert hasNotCalledProcess() {
    final Set<String> calledProcesses =
        getCalledProcessRecords().stream()
            .map(process -> process.getValue().getBpmnProcessId())
            .collect(Collectors.toSet());

    assertThat(!calledProcesses.isEmpty())
        .withFailMessage(
            "A process was called from this process, distinct called processes are: %s",
            calledProcesses)
        .isFalse();
    return this;
  }
```

You will have to swap the processes in the failure message of your unit test.

Wolgo replied:
Ah, Collectors.toSet() removes the need for distinct. I still gotta get more used to streams in Java. 
Thanks for the reviewing!

remcowesterhoud reviewed (changes requested):
🚀 Great job, there is 1 more minor thing. After this I can merge the PR 🎉

remcowesterhoud reviewed (approved):
All good! Thanks for the contribution 🏆 And good luck with the new job and house 🙂

remcowesterhoud commented:
/backport
